wyzesense2mqtt.py

```diff
@@ -12,169 +10,266 @@
+    MQTT_CLIENT.on_connect = on_connect
+    MQTT_CLIENT.on_disconnect = on_disconnect
+    MQTT_CLIENT.on_message = on_message
+
+    # Connect to MQTT
+    LOGGER.info("Attempting to open connection to MQTT host {0}".format(CONFIG['mqtt_host']))
+    MQTT_CLIENT.connect(CONFIG['mqtt_host'], port = CONFIG['mqtt_port'], keepalive = CONFIG['mqtt_keepalive'])
+
+# Initialize USB dongle
+@retry(wait_exponential_multiplier=1000, wait_exponential_max=10000)
+def init_wyzesense_dongle():
+    global WYZESENSE_DONGLE, CONFIG
+    if CONFIG['usb_dongle'].lower() == "auto": 
+        df = subprocess.check_output(["ls", "-la", "/sys/class/hidraw"]).decode("utf-8").lower()
+        for l in df.split("\n"):
+            if ("e024" in l and "1a86" in l):
+                for w in l.split(" "):
+                    if ("hidraw" in w):
+                        CONFIG['usb_dongle'] = "/dev/%s" % w
+                        break
+
+    LOGGER.info("Openning connection to dongle {0}".format(CONFIG['usb_dongle']))
+    try:
+        WYZESENSE_DONGLE = wyzesense.Open(CONFIG['usb_dongle'], on_event)
+        LOGGER.debug("  MAC: {0}, VER: {1}, ENR: {2}".format(WYZESENSE_DONGLE.MAC, WYZESENSE_DONGLE.Version, WYZESENSE_DONGLE.ENR))
+    except IOError:
+        LOGGER.warning("No device found on path {0}".format(CONFIG['usb_dongle']))
+
+# Initialize sensor configuration
+def init_sensors():
+    global SENSORS
+    LOGGER.debug("Reading sensors configuration...")
+    if os.path.isfile(SENSORS_CONFIG_FILE):
+        SENSORS = read_yaml_file(SENSORS_CONFIG_FILE)
     else:
-        _LOGGER.debug("Empty Scan")
-
-# Process message to remove device
-def on_message_remove(client, userdata, msg):
-    _LOGGER.info("In on_message_remove: {0}".format(msg.payload.decode()))
-    sensor_mac = msg.payload
-
-    ws.Delete(sensor_mac)
-
-    clear_retained_mqtt_topics(sensor_mac)
-
-# Send HASS discovery topics to MQTT
-def send_discovery_topics(sensor_mac, sensor_type):
-    _LOGGER.info("Publishing discovery topics")
+        LOGGER.info("No sensors config file found.")
+
+    for sensor_mac in SENSORS:
+        if valid_sensor_mac(sensor_mac):
+            send_discovery_topics(sensor_mac)
+
+    # Check config against linked sensors
+    try:
+        result = WYZESENSE_DONGLE.List()
+        LOGGER.debug("Linked sensors: {0}".format(result))
+        if result:
+            for sensor_mac in result:
+                if valid_sensor_mac(sensor_mac):
+                    if SENSORS.get(sensor_mac) is None:
+                        add_sensor_to_config(sensor_mac)
+                        send_discovery_topics(sensor_mac)
+        else:
+            LOGGER.warning("Sensor list failed with result: {0}".format(result))
+    except:
+        pass
+
+# Validate sensor MAC
+def valid_sensor_mac(sensor_mac):
+    if len(sensor_mac) == 8 and sensor_mac != "00000000" and sensor_mac != "\0\0\0\0\0\0\0\0" and sensor_mac != "\x00\x00\x00\x00\x00\x00\x00\x00":
```

Comment:
This is 100% a me thing, and it might be "unpythonic", but I like surrounding each condition in a multi-condition `if` in parentheses to make it more clear where one ends and the next begins. ex:

`if (len(sensor_mac) == 8) and ...`

Reply:
No idea on "pythonic" or not, but easy enough to do, and done in next commit.